### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by ereteog**
> related #https://github.com/threatgrid/iroh/issues/4381

This PR introduces Authentication in the Elasticsearch connection.
A generic schema is proposed to provide the authentication parameters, and the main methods are implemented.

### CODE REVIEW

File: test/ductile/auth_test.clj [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+(deftest basic-auth-test
+  (is (= {:basic-auth ["the-login" "the-pwd"]}
+         (sut/basic-auth {:user "the-login" :pwd "the-pwd"}))))
+
+(deftest oauth-token-test
+  (is (= {:oauth-token "any-token"}
+         (sut/oauth-token {:token "any-token"}))))
+
+(deftest bearer-test
+  (is (= {:oauth-token "Bearer any-token"}
```
Comment (vemv) [inline test/ductile/auth_test.clj:20]: Here you can also assert that "Bearer any-token" -> "Bearer any-token", i.e. existing `Bearer ` is omitted

File: test/ductile/conn_test.clj [hunk 2]
```diff
@@ -15,63 +15,84 @@
...
+           (sut/make-http-opts nil opts [:param-1])))))
+
+(deftest testing-auth-params
+  (testing "auth options should be properly prepared and integrated in http params."
+    (let [test-auth (fn [auth-params expected]
+                      (let [{:keys [auth] :as conn}
+                            (sut/connect {:host "127.0.0.1"
+                                          :port 9200
+                                          :auth auth-params})
+                            auth-http-opts (-> (sut/make-http-opts conn {})
```
Comment (vemv) [inline test/ductile/conn_test.clj:80]: You could also exercise that wrong creds result in a connection not being granted

**Review by vemv — APPROVED** [review]
LGTM, great work!

**Review by agzam — APPROVED** [review]
Very nice! 

Have you ever thought about using Hiccup and generating doc/*.html, instead of messing with it directly? Feels a bit strange not to want to use s-exps instead (when that's possible with virtually zero overhead).

**Comment by ereteog** [comment]
> Have you ever thought about using Hiccup and generating doc/*.html, instead of messing with it directly? Feels a bit strange not to want to use s-exps instead (when that's possible with virtually zero overhead).

I only written `intro.md` and used [codox](https://github.com/weavejester/codox) to generate the `doc/*html`.

File: test/ductile/conn_test.clj (hunk 2 above)
Comment (ereteog) [reply test/ductile/conn_test.clj:80]: I [added an integration test](https://github.com/threatgrid/ductile/pull/9/commits/946196660b846db555fbc0518060261e8511cc3f) with authorized and unauthorized requests with missing auth, basic auth and API key.

**Pull request merged**